tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,225 @@
+#           if (names_opt.has_value()) {
+#                         ~~~~~~~~~~~ <-- Unwrapping optional shell
+#               const c10::IValue names_opt_in = names_opt.value();
+#               const c10::List<c10::IValue> names_list_in = names_opt_in.toList();
+#
+#               std::vector<at::Dimname> names_vec;
+#               for (c10::IValue names_elem: names_list_in) {
+#                                ~~~~~~~~~~~~~~~~~~~~~~~~~ <-- Unrolling list, then convert elements one by one.
+#                   at::Dimname names_base = names_elem.to<at::Dimname>();
+#                   names_vec.push_back(names_base);
+#               }
+#               at::ArrayRef<at::Dimname> names_list_out(names_vec);
+#
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>(names_list_out);
+#           } else {
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>();
+#           }
+#       ```
+#    - ScalarType? dtype (similarly for the rest of the arguments)
+#       ```cpp
+#           c10::optional<c10::IValue> dtype_opt = (std::move(peek(stack, 2, 7))).toOptional<c10::IValue>();
+#           c10::optional<at::ScalarType> dtype_opt_out;
+#           if (dtype_opt.has_value()) {
+#               const c10::IValue dtype_opt_in = dtype_opt.value();
+#               at::ScalarType dtype_base = dtype_opt_in.to<at::ScalarType>();
+#                                                        ~~~~~~~~~~~~~~~~~~~~ <-- For base types, convert ivalue to it
+#                                                                                 directly using ".to<T>()" API.
+#               dtype_opt_out = c10::optional<at::ScalarType>(dtype_base);
+#           } else {
+#               dtype_opt_out = c10::optional<at::ScalarType>();
+#           }
+#       ```
+#
+# - Unboxed Kernel Call
+#   ```cpp
+#       auto result_ = torch::empty(
+#           size_list_out,
+#           names_opt_out,
+#           options,
+#           memory_format_opt_out
+#       );
+#   ```
+#
+# - Push Result Back to Stack
+#   ```cpp
+#       drop(stack, 7);
+#       pack(stack, std::move(result_));
+#   ```
+
+# Convert all the arguments in a NativeFunction to C++ code, including TensorOptions.
+def convert_arguments(args: List[Argument]) -> Tuple[List[Expr], List[str]]:
+    argument_str = "(std::move(peek(stack, {pos}, {args_num})))"
+    expr_list = []
+    code_list = []
+    for i, arg in enumerate(args):
+        ivalue_str = argument_str.format(pos=i, args_num=len(args))
+        expr, code = argumenttype_ivalue_convert(arg.type, ivalue_str, arg.name, mutable=arg.is_write)
+        expr_list.append(expr)
+        code_list.extend(code)
+
+    return expr_list, code_list
+
+
+# Take an argument in JIT type format, returns the C++ code to convert an ivalue from stack to corresponding C++ type.
+def argumenttype_ivalue_convert(t: Type, ival: str, arg_name: ArgName, *, mutable: bool = False) -> Tuple[
+    Expr, List[str]]:
```

Comment:
I think it's a little confusing that this function takes in an `ival: str` and a separate `arg_name: ArgName`. I see now that `ival` corresponds to C++ code with the transformation to grab the ivalue to the stack, and `arg_name` is the name of the original variable in the native function schema.

I think there are a couple ways we could make that clearer. One would be to assign the ivalues to temporary variables that we can refer to directly in the codegen, like:
```
TORCH_API void add_Tensor(Stack & stack) {
    auto self_ivalue = std::move(peek(stack, 0, 3));
    at::Tensor self_base = self_ivalue.to<at::Tensor>();
    ...
}
```

Then in the codegen you could have two functions:
- one for unwrapping the ivalue (returning both the code-string to unwrap, and a `Binding` corresponding to the unwrapped value)
- another for unboxing the individual IValue (aka this function). Then this function can just take in the Binding and nothing else (it can figure out `Type` and `mutable` from `Binding.argument`)

------

Another related idea: what do you think of manually writing the functions that unbox each type directly in the template, like `unbox_TensorList`, `unbox_OptionalTensorList`, etc? That seems nice because any function that lives directly in C++ is way easier to reason about than one that's indirectly generated from python codegen. Then this function becomes a lot simpler - Given a `Binding` with some `name`, the code string it spits out is just a single function call (and you don't have to deal with doing any of the recursive calls that you have here for container types, etc).

Reply:
> manually writing the functions that unbox each type directly in the template, like unbox_TensorList, unbox_OptionalTensorList, etc? 

That is a bit overwhelming since we have more than 20 types in native_functions.yaml, also it cannot handle a type that is not known from native_functions.yaml